### PR TITLE
Added code

### PR DESCRIPTION
**Pull request by calnix**
I implemented margin based borrowing; similar to how FX brokers operate. This made sense to me from:
1. the perspective of lending, (as brokers will never be in a position of risk and will margin call you at some level before)
2. having a buffer would prove useful in volatile markets combined with network congestion.

Please note, V2 contracts are for stretch goals:
1. Most tokens have 18 decimals, but not all. Code another Collateralized Vault where the underlying is USDC.
2. Using mock WETH and DAI contracts gets you a long way, but you can't always do that. Remove the mint() function and use the foundry cheatcodes to directly edit the registry that stores your balance of each.

### CODE REVIEW

File: src/Vault.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@dev To set collateralization level of debt engine
+    ///@notice To allow for margin call at defined level, preventing Vault from bearing risk in fast moving markets
+    uint public immutable bufferDenominator; 
+    uint public immutable bufferNumerator;
```
Comment (alcueca) [inline src/Vault.sol:60]: You could just use one fixed point number for this, it would be simpler.

File: src/Vault.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    ///@dev Returns decimal places of asset prices as dictated by Chainlink Oracle
+    uint public immutable decimals;
```
Comment (alcueca) [inline src/Vault.sol:63]: Since you are always going to do `10**decimals` you could store this as a scalar factor and do that power operation only in the constructor.

File: src/Vault.sol [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for margined colleteralization colculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint maxDebt = getMaxDebt(msg.sender);
```
Comment (alcueca) [inline src/Vault.sol:90]: You have to functions to check collateralization, `getMaxDebt` and `getCollateralRequired`, which you use only to get a yes/no answer on whether a vault is sufficiently collateralized. If you make this precondition into a postcondition, and check after `debts[msg.sender] += debtAmount;` you can use the same collateralization check as in `withdraw`.

Alternatively, you can make the collateralization check in `withdraw` a post-condition, and use `getMaxDebt` for the check. But you use `getCollateralRequired` in `liquidate`, so I guess that's the good one.

In short, you can remove one of the two.

For color, [this is how MakerDAO does it](https://github.com/makerdao/dss/blob/17187f7d47be2f4c71d218785e1155474bbafe8a/src/vat.sol#L163).

I think Dan Robinson told me of this one when coding Yield v1 :D

File: src/Vault.sol [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        if (collateralRequired > deposits[user]){
+            emit Liquidation(address(collateral), address(debt), user, debts[user], deposits[user]); 
+            delete deposits[user];
+            delete debts[user];
+        }
+    }
+
+    ///@notice Price is returned as integer extending over its decimal places
+    ///@dev Get current market price of Collateral/Debt asset from Chainlink
+    function get_daiETH_Price() public view returns(uint) {
```
Comment (alcueca) [inline src/Vault.sol:160]: When you collapse `getMaxDebt` and `getCollateralRequired` into one, you can just add the oracle call in that function, and remove the separate `get_daiETH_Price`.

File: src/VaultV2.sol [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for colleteralization colculation
```
Comment (alcueca) [inline src/VaultV2.sol:91]:
```suggestion
    ///@dev See getMaxDebt() for collateralization calculation
```

**Review by alcueca — CHANGES_REQUESTED** [review]
This is a partial review, just to simplify the code a bit first.

File: src/Vault.sol [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    mapping (address => uint) public deposits; 
+
+    ///@dev Vault records debt holdings of each user
+    mapping (address => uint) public debts;  
+
+
+    ///@notice ERC20 interface specifying token contract functions
+    ///@dev For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
+    IERC20 public immutable collateral;    
+    IERC20 public immutable debt;
```
Comment (devtooligan) [inline src/Vault.sol:20]: i dont love this varname. debt sounds like a number.  yield calls this `base`

File: src/Vault.sol [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@dev To set collateralization level of debt engine
+    ///@notice To allow for margin call at defined level, preventing Vault from bearing risk in fast moving markets
+    uint public immutable bufferDenominator;
```
Comment (devtooligan) [inline src/Vault.sol:59]: this feels like something that we would might want to change later, but then u have to use a storage var which kinda sucks for gas usage.  if u do change this to a storage var i would use a ray instead of storing both numerator and denominator. in fact i would probably do that here anyways.  even a ray might be too much, how many digits of precision do u really need?  🤔

File: src/Vault.sol [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        bufferNumerator = bufferNumerator_;
+        bufferDenominator = bufferDenominator_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");
```
Comment (devtooligan) [inline src/Vault.sol:82]: eh, this feels a bit overkill tbh.  @alcueca ?

File: src/Vault.sol [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Get maximum debt asset borrowable, calculated based on collateralization level
+    ///@dev Collateralization level: bufferNumerator/bufferDenominator (e.g. 8/10 -> 80%) | Think of this as margining
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
```
Comment (devtooligan) [inline src/Vault.sol:103]: get in the habit of multiplying first then dividing after, this wei u wont lose precision

File: src/Vault.sol [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {
```
Comment (devtooligan) [inline src/Vault.sol:123]: i appreciate u trying to have a descriptive name, but this var name also seems confusing to me.  i think `amount` is fine or else `withdrawAmount`

File: src/Vault.sol [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
+        
+        if (debts[msg.sender] > 0){
```
Comment (devtooligan) [inline src/Vault.sol:125]:
```suggestion
        uint debt = debts[msg.sender]; // caching this saves an SLOAD
        uint deposit = deposits[msg.sender]; // caching this saves an SLOAD
        uint collateralRequired;
        if (debts[msg.sender] > 0){
            collateralRequired = getCollateralRequired(debt);
        }
        require(withdrawAmount <= deposit - collateralRequired, "whatever");
        
        deposit[msg.sender] = deposit - withdrawAmount;
```

File: src/Vault.sol [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            uint collateralRequired = getCollateralRequired(debts[msg.sender]);
+            uint spareDeposit = deposits[msg.sender] - collateralRequired;
+            require(collateralAmount < spareDeposit, "Collateral unavailable!");
+        }
+        
+        deposits[msg.sender] -= collateralAmount;
```
Comment (devtooligan) [inline src/Vault.sol:131]:
```suggestion
        uint debt = debts[msg.sender]; // caching this saves an SLOAD
        uint deposit = deposits[msg.sender]; // caching this saves an SLOAD
        uint collateralRequired;
        if (debts[msg.sender] > 0){
            collateralRequired = getCollateralRequired(debt);
        }
        require(withdrawAmount <= deposit - collateralRequired, "whatever");
        
        deposit[msg.sender] = deposit - withdrawAmount;
```

File: src/Vault.sol [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
+        uint maxDebt = (availableCollateral*10**decimals / get_daiETH_Price());
+        return maxDebt;
+    }
+
+    ///@notice Users repay their debt; in debt asset terms
+    ///@dev This covers partial and full repayment scenarioes
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
```
Comment (devtooligan) [inline src/Vault.sol:111]: @alcueca in the past we've always been sticklers about typiing out `uint256` instead of just `uint`  -- tbh lately ive been feeling open to `uint` -- wdyt?

File: src/Vault.sol [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        bool sent = collateral.transfer(msg.sender, collateralAmount);
+        require(sent, "Withdraw failed!");
+        emit Withdraw(address(collateral), msg.sender, collateralAmount);            
+    }
+
+
+    ///@dev Calculates collateral required to support existing debt position at current market prices 
+    ///@param debtAmount Amount of debt asset to support
+    function getCollateralRequired(uint debtAmount) public view returns(uint) {
+        uint baseCollateralRequired = debtAmount*get_daiETH_Price() / (10**decimals);
```
Comment (devtooligan) [inline src/Vault.sol:141]: this looks weird to me without spaces, i would do:

```
    ... debtAmount * get_daiEth_Price()
```

i guess technically the [style guide](https://docs.soliditylang.org/en/v0.8.13/style-guide.html#other-recommendations) gives some leeway to group things without spaces, so this is more of a nit i guess
> Operators with a higher priority than others can exclude surrounding whitespace in order to denote precedence. This is meant to allow for improved readability for complex statements. You should always use the same amount of whitespace on either side of an operator:

File: src/Vault.sol [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        uint bufferCollateralRequired = (baseCollateralRequired/bufferNumerator)*bufferDenominator;
+        return bufferCollateralRequired;
```
Comment (devtooligan) [inline src/Vault.sol:143]: you could just return the above instead of setting it to a var first

alternatively, you could use a named return.  that means above in the func sig you would write:
```
   ... public view returns(uint bufferCollateralRequired) {...
```

and so here you wouldnt even need a return (u also wouldnt need to use the type in the line above)

File: src/Vault.sol [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Get maximum debt asset borrowable, calculated based on collateralization level
+    ///@dev Collateralization level: bufferNumerator/bufferDenominator (e.g. 8/10 -> 80%) | Think of this as margining
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
+        uint maxDebt = (availableCollateral*10**decimals / get_daiETH_Price());
```
Comment (devtooligan) [inline src/Vault.sol:104]: idt ur math works if base decimals are different from ilk decimals

File: src/Vault.sol [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+    ///@dev Vault records collateral deposits of each user
+    mapping (address => uint) public deposits; 
+
+    ///@dev Vault records debt holdings of each user
+    mapping (address => uint) public debts;  
+
+
+    ///@notice ERC20 interface specifying token contract functions
+    ///@dev For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
```
Comment (devtooligan) [inline src/Vault.sol:18]: i know its tough to think of @dev notes but u dont need to provide basic info about common syntax lol

File: src/Vault.sol (hunk 4 above)
Comment (devtooligan) [inline src/Vault.sol:160]: why are we using daiEth price?? im confused.  i thought this vault can take any 2 erc 20 tokens....

File: src/Vault.sol (hunk 14 above)
Comment (devtooligan) [inline src/Vault.sol:141]: idt ur math handles base / ilk with differing decimals correctly

Comment (devtooligan) [reply src/Vault.sol:141]: try an example with a base of 6 deci and ilk of 18

File: src/VaultV2.sol [hunk 18]
```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (devtooligan) [inline src/VaultV2.sol:1]: not sure what this VaultV2 is....didnt review

File: test/VaultTest.t.sol [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+    }
+
+    function testOnlyOwnerCanCallLiquidate() public {
+        console2.log("Only Owner of contract can call liquidate -onlyOwner modifier");
+        vm.prank(user);
+        vm.expectRevert("Ownable: caller is not the owner");
+        vault.liquidation(user);
+    }
+}
+
```
Comment (devtooligan) [inline test/VaultTest.t.sol:256]: tests look pretty great.  i think uve covered all the states. i like the organization of the tests too.   well done

the only thing I would add is I would include some tests that the rate changes mid way,  so like -- borrow half of max, change rate, expect revert on borrowing the other half (because of rate change) borrow less and it works

File: test/VaultTestV2.t.sol [hunk 20]
```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (devtooligan) [inline test/VaultTestV2.t.sol:1]: im still confused about this v2 thing so did not review these

File: src/Vault.sol (hunk 13 above)
Comment (calnix) [reply src/Vault.sol:111]: I recall reading that gas-wise they were the same  and there no savings by specifying 256 or not.

So it seemed to me to be sensible to not bother - unless I've read wrong

File: src/VaultV2.sol (hunk 18 above)
Comment (calnix) [inline src/VaultV2.sol:1]: I approached by tackling the DAI/ETH example as a base before updating to take in any two ERC20 tokens - potentially with different decimals; as I commented at the top.

Vault.sol -> specific to DAI/ETH 
VaultV2.sol -> extended for generality as per stretch goals

File: test/VaultTest.t.sol (hunk 19 above)
Comment (calnix) [reply test/VaultTest.t.sol:256]: Oooh, that's a great idea. Thanks!

File: src/Vault.sol (hunk 4 above)
Comment (calnix) [reply src/Vault.sol:160]: Vault.sol is specific to the DAI/ETH example in the issue, while VaultV2.sol is the general implementation where one could have tokens of differing decimal precision.

Although tbh, I did not change the name of the price function in V2 - so that's definitely an error on my part.

File: src/Vault.sol (hunk 17 above)
Comment (calnix) [reply src/Vault.sol:18]: lol I kept that around for myself tbh, so it would stick in my brain as a habit. will remove ;)

File: src/Vault.sol (hunk 9 above)
Comment (calnix) [reply src/Vault.sol:103]: lol nice pun - yup its something i need to be concussions about

File: src/Vault.sol (hunk 17 above)
Comment (devtooligan) [reply src/Vault.sol:18]: fair enough!

File: src/Vault.sol (hunk 7 above)
Comment (calnix) [reply src/Vault.sol:59]: in my imagining, these levels are usually set in some integer value btw 70-80%, at least for retail fx brokers - so not much precision would be required. 

as for changing it, I'm not sure. brokers rarely modify margin call levels, but perhaps my analogy itself might be off-tangent, you'll have to correct me 😕

Comment (devtooligan) [reply src/Vault.sol:59]: if not much precision is required, then this is even more of an argument to use a fp number.  as for changing it, i see u what u mean about margin calls.  at yield we have different collateralization rates depending on the base and collateral -- and sometimes we make mistakes or change our mind about the collateralization ratio (also sometimes external factors could arise that might make us want to change)  so thats where i was coming from.

for this exercise u can leave it as immutable, but i would suggest a fp number instead of 2 integers, numerator/denominator

File: src/Vault.sol (hunk 14 above)
Comment (calnix) [reply src/Vault.sol:141]: I do that in V2.
V2 being updated as per the stretch goals to take in any two ERC20 tokens - potentially with different decimals

File: src/Vault.sol (hunk 4 above)
Comment (calnix) [reply src/Vault.sol:160]: ahhh thats much more elegant, thanks

File: src/Vault.sol (hunk 9 above)
Comment (calnix) [reply src/Vault.sol:103]: removed this function as advised by Alberto

File: src/Vault.sol (hunk 12 above)
Comment (calnix) [reply src/Vault.sol:131]: Ah i thought i was saving gas with direct referencing and not explicitly storing into memory. Any suggestions on where is a good place to get started on learning the assembly stuff that encompass SLOAD et al?

File: src/Vault.sol (hunk 4 above)
Comment (devtooligan) [reply src/Vault.sol:160]: So sorry i just read the requirements again.  It was different than what i was remembering.  I will re review and cleanup my review tomorrow. I think v2 must have the math im looking for

File: src/Vault.sol (hunk 12 above)
Comment (devtooligan) [reply src/Vault.sol:131]: Use the debugger.  Walk through some code. 

But tbh u can just remember any time u acess a storage slot its an sload. Cold (first) sload is 2100 gas.  Warm 2nd is 100. New local mem is 6.  

Also remember SLOAD is for 1 word (32 bytes) so if u line up your storage vars correctly u could pack 2x 128 or multiple smaller sized  into a single SLOAD

File: src/Vault.sol (hunk 13 above)
Comment (devtooligan) [reply src/Vault.sol:111]: Its just a style guide thing

File: src/Vault.sol (hunk 4 above)
Comment (calnix) [reply src/Vault.sol:160]: Don't worry, to make it easier for both Alberto and you, I'm going to flatten both versions into a single generic version and upload as a new branch(addressing the comments provided so far). 

Cleaner and less confusion that way :)
